lib/pacemaker/pcmk_cluster_queries.c

```diff
@@ -406,19 +431,66 @@ pcmk__list_nodes(pcmk__output_t *out, gboolean BASH_EXPORT)
         return pcmk_legacy2rc(rc);
     }
 
-    rc = the_cib->cmds->query(the_cib, NULL, &output,
+    rc = the_cib->cmds->query(the_cib, NULL, &xml_node,
                               cib_scope_local | cib_sync_call);
     if (rc == pcmk_ok) {
-        out->message(out, "crmadmin-node-list", output, BASH_EXPORT);
-        free_xml(output);
+        xmlNode *node = NULL;
+        xmlNode *nodes = get_object_root(XML_CIB_TAG_NODES, xml_node);
+        struct node_data data = {
+            .out = out,
+            .found = 0,
+            .BASH_EXPORT = BASH_EXPORT
+        };
+
+        out->begin_list(out, NULL, NULL, "nodes");
+
+        if (!pcmk__str_empty(node_types) && strstr(node_types, "all")) {
+            node_types = NULL;
+        }
+
+        if (pcmk__str_empty(node_types) || strstr(node_types, "cluster")) {
+            for (node = first_named_child(nodes, XML_CIB_TAG_NODE); node != NULL;
+                 node = crm_next_same_xml(node)) {
+                const char *node_type = crm_element_value(node, XML_ATTR_TYPE);
+                if (node_type == NULL) {
```

Comment:
I think we also need to check for type="member". I don't think it's possible for the cluster to create that anymore, but we do have it in old scheduler regression test inputs (e.g. cts/scheduler/1360.xml)

Alternatively you could replace this for loop with a crm_foreach_xpath_result() using PCMK__XP_MEMBER_NODE_CONFIG